feat(server): add ability to change builds concurrency (reopened)

NoeGrangier opened the pull request:
# Summary

This PR adds the ability to configure concurrent builds in Dokploy.
Users can now change the number of simultaneous builds from the default of 1 to up to 20 through the settings interface.

**Before**: All deployments (local + all remote servers) shared one queue with a single concurrency limit set to one by default.

**After**: Each remote server now has its own dedicated deployment queue and worker with independent concurrency settings.

This feature will be quite usefull for my team and I as we are using dokploy more and more and are getting a bit limited when a developper has to wait for others apps to build before its own.

We've been loving Dokploy so far and I'm very happy to try to help :)

Feel free to share what you think about it !

## Changes

- **Database**: Added `buildsConcurrency` column to user table (default: 1)
- **Backend**: Added API endpoints to get/update build concurrency settings
- **Frontend**: Added modal dialog in **Settings** → **Web Server** → **Server** → **Change builds concurrency** or for remote servers **Remote servers** → **Actions** → **Change builds concurrency** 
- **Workers**: Created a map of workers (one for each remote server) with each their own concurrency setting

## Technical details
I've made the deployment workers as a map of singleton to make it available to the Trpc endpoints.
However, I'm not sure about the implementation but it's one I could get working reliably.

## How to Use

1. Go to **Settings** → **Web Server** → **Server** → **Change builds concurrency** or **Remote servers** → **Actions** → **Change builds concurrency**
2. Enter number (1-20) and save

## Resource Warning

The UI warns that each concurrent build needs ~2 vCPUs + 4GB RAM.
It may be update for debate, I took this from the warning about builders ressource consumption.

## Screenshots

<img width="1696" alt="image" src="https://github.com/user-attachments/assets/25317802-5b9f-47ac-8b69-0babdb4f7719" />

<img width="1179" alt="image" src="https://github.com/user-attachments/assets/2d5c369c-e1a4-4726-9c71-9750d7e9daf7" />

## Related issues
Closes #1283
Partially closes #2000 
Mentionned here: #877

NoeGrangier committed:
feat(server): add ability to change builds concurrency

NoeGrangier committed:
feat(server): Added queue for each remote server

NoeGrangier committed:
Add migrations

Siumauricio committed:
Remove obsolete SQL migration and associated journal entries for buildsConcurrency in user_temp and server tables.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

Siumauricio committed:
feat(database): add buildsConcurrency column to user_temp and server tables

- Introduced a new column "buildsConcurrency" of type integer with a default value of 1 in both "user_temp" and "server" tables to enhance concurrency management.
- Updated journal and snapshot files to reflect these schema changes, ensuring consistency across the database.

Siumauricio committed:
test(server): add buildsConcurrency property to baseAdmin user object in update-server-config test

Siumauricio committed:
Remove obsolete SQL migrations for buildsConcurrency in user_temp and server tables.

Siumauricio committed:
Refactor deployment worker initialization to use buildsConcurrency from admin user object

- Removed unused User type and getWorkersMap import.
- Simplified concurrency handling by directly assigning the buildsConcurrency value from the admin user object, defaulting to 1 if not available.
- Improved code readability and maintainability in server.ts.

Siumauricio commented:
When saving the values I'm getting this error: Deployment worker is not available in getBuildsConcurrency

<img width="1263" alt="image" src="https://github.com/user-attachments/assets/3b161e02-690e-4507-a21e-a8a23459abce" />

NoeGrangier committed:
fix: Fixed default worker value not being correct

NoeGrangier commented:
@Siumauricio My bad, the default key for the main server queue was wrong, I've fixed it.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

Siumauricio committed:
feat(database): add buildsConcurrency column to user_temp and server tables

- Added "buildsConcurrency" column with a default value of 1 to enhance concurrency management.
- Updated journal and snapshot files to reflect the new schema changes.

Marukome0743 commented:
Could you resolve merge conflicts?

NoeGrangier committed:
Merge migration

NoeGrangier committed:
migrations

NoeGrangier commented:
Hey @Siumauricio, any news on this one ? Our team is really looking forward to it :)

Marukome0743 commented:
Thank you so much!
Could you resolve merge conflicts?

Siumauricio committed:
Remove deprecated SQL migration files and corresponding snapshot metadata for buildsConcurrency and isolatedDeploymentsVolume columns.

NoeGrangier committed:
Added SQL migration after rebase

NoeGrangier commented:
@Marukome0743 Done :)

ardasevinc commented:
it'd be awesome to get this merged. building one at a time is so archaic. love dokploy and we need this!

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

Siumauricio committed:
Merge branch 'canary' into feat/deployment-queues

Siumauricio committed:
Remove deprecated SQL migration files and update journal to reflect changes. The `0106_round_scrambler.sql` and `0106_snapshot.json` files have been deleted as they are no longer needed.

Siumauricio committed:
fix(json): ensure newline at end of file in 0099_snapshot.json

Siumauricio committed:
feat(snapshot): add new snapshot files for application schema updates and ensure proper formatting

Siumauricio committed:
feat(journal): add new entry for version 7 with tag '0106_purple_maggott'

Siumauricio committed:
feat(database): add buildsConcurrency column to user_temp and server tables, and update journal for version 7

AminDhouib commented:
Getting notified of your commits @Siumauricio and it's getting me excited to get this merged. Very excited to get this for my team!

Siumauricio commented:
Thank you for your interest. I am looking into this, as I have wanted to change this for some time, but I am considering removing bullmq and redis.

Because, looking at the current situation and the problems that currently exist, such as this one https://github.com/Dokploy/dokploy/issues/2106, and also because we cannot cancel ongoing builds (there are many issues surrounding this), 

I have been thinking about having a much simpler system with queues in memory without redis

For example, if you are doing a build and your server freezes due to the Docker build, bullmq will keep trying again and again due to its mechanism of saving jobs and retrying, and you are basically in an infinite loop.

So I was thinking, if we used a memory queue, if a freeze occurs, at least you can recover, but with the downside that your deployments that were in the queue will be lost due to the restart. However, I prefer that the system can recover and lose the deployments. I'm not sure what you think yet.

AminDhouib commented:
@Siumauricio thinking for now, you could allow the multiple concurrency feature. Make it only be able to run 1 by default as you do. Do not advertise that Dokploy can make concurrent builds but put it as an option. If somebody wants to select anything above 1, they would get a warning highlighting the major problem that may occur and they are at risk of losing their web server and would be a good idea to backup.

Basically, it's better to provide the option than not as soon as possible since it seems to be a highly requested feature. As an example, I have a 32GB RAM + 16vCPU setup for Dokploy. I would rather have the option and allow up to 4 concurrent builds than being limited by 1 despite my hardware being absolutely capable of doing 4-8 concurrent builds very easily.

Hope this perspective helps!

NoeGrangier commented:
Agree, I think the in-memory queue would have huge benefits (mainly the ability to cancel an ongoing build).
But I think that while waiting for this to be implemented, allowing us to have multiple build concurently now would have huge benefits for our team as we're having more and more projects with multiple builds a day.

Hope it gets merged for now and reworked later :)

ardasevinc commented:
would love to help whatever we can

mostafa-binesh commented:
+1

daniel27si commented:
+1

saidbenmoumen commented:
+1

baautitz commented:
+1

mikhaelr commented:
Any update on this? This would be really valuable when using a single Dokploy instance to manage multiple projects, especially for agencies with lots of clients. Currently, the experience gets significantly worse as the number of projects grows and deployments become more frequent. Do you have an idea of when this might get merged? Thanks a lot for the work on this

matheusbento commented:
+1

fredEbho commented:
This is highly needed! +1

austinbiggs commented:
++ to all of these messages, this feature would be so beneficial. I have multiple services that I'm deploying for a single distributed system. An API, MCP, Docs, and a CMS all of which tend to receive changes in larger PRs. All of which are separate deployed services in Dokploy.

Dokploy is the perfect alternative to using Vercel for builds and this is the last puzzle piece. Concurrent builds would be a tremendous improvement worthy of praise.

We're all here to help, if the core team would please consider and communicate, we the community would be ecstatic.

Thank you for reading 🙏

weibeu commented:
Really respect the work of developers here and thanks to keep this really great tool opensource. This has really really helped our teams streamline their deployment lifecycle.

I understand it takes a lot of efforts, to maintain such projects. However, still bumping this up to highlight that this issue is getting a sort of blocker / bottleneck now.

We are happy to contribute wherever necessary and get this pushed.

mostafa-binesh commented:
Why this issue is not getting any attention from contributers?

ajnart commented:
Thank you @AminDhouib ! I am planning to do the same style of community edition or contribute to yours since @Siumauricio has stopped accepting contributions

Siumauricio commented:
Thank you for the work you put into this PR and for everyone's patience — this has been one of the most requested features and this PR helped shape the direction we ended up taking.

I'm closing this PR because we've now shipped a native implementation of concurrent builds, with a different architecture than what was proposed here:

What changed:

1. Per-server queues: Previously, Dokploy had a single global queue shared across all deployments (local + remote servers). Now each server has its own independent queue with its own concurrency settings, so builds on one server no longer block builds on another.
2. Redis/BullMQ removed: We replaced the Redis-backed queue with an in-memory queue. As I mentioned earlier in this thread, BullMQ's retry mechanism could put a frozen server into an infinite loop, and canceling ongoing builds was practically impossible. With the in-memory queue, the system can always recover from a freeze. As a bonus, Dokploy now depends on one less container to run, making the whole setup lighter and easier to operate.
3. Concurrency limits: OSS supports up to 2 concurrent builds per server, and Enterprise supports unlimited concurrent builds.

Thanks again @NoeGrangier for pushing this forward. If you run into any issues with the new implementation, please open an issue and we'll take a look 🙏

https://github.com/Dokploy/dokploy/releases/tag/v0.29.9

https://docs.dokploy.com/docs/core/concurrent-builds

ajnart commented:
No way what a surprise! 
Once again @Siumauricio ignored the community to steal ideas and implement them as their own, including that sweet enterprise limitation on top of it

RDeluxe commented:
Sadly, I am not surprised as I feared that this PR was blocked due to business reasons. 

Still, this is a huge disappointment. I know that times are tough and that making money with OSS is harder than ever, but I'm not sure that's the right way of doing so: concurrent build is a core feature, limiting it that much feels like a bad joke. 

I feel like we won't have a choice but to maintain of light fork, disappointing when my team has been actively contributing here. 

Regards

Siumauricio commented:
Hey @ajnart @RDeluxe — I hear the frustration, and I want to be transparent about where we landed.

During our Q3 planning today, we discussed the concurrent builds feature specifically. We've decided to make it fully open to everyone, with no Enterprise-only limits. We see concurrent builds as a core platform capability, and core capabilities belong in the open-source version — that's the open core model we want to follow.

Balancing sustainability with keeping Dokploy genuinely useful as OSS isn't always easy, and we won't always get the line right on the first try. This thread was part of what pushed the discussion. The change will ship in an upcoming release.

Thanks you!

ardasevinc commented:
> Hey @ajnart @RDeluxe — I hear the frustration, and I want to be transparent about where we landed.
> 
> 
> 
> During our Q3 planning today, we discussed the concurrent builds feature specifically. We've decided to make it fully open to everyone, with no Enterprise-only limits. We see concurrent builds as a core platform capability, and core capabilities belong in the open-source version — that's the open core model we want to follow.
> 
> 
> 
> Balancing sustainability with keeping Dokploy genuinely useful as OSS isn't always easy, and we won't always get the line right on the first try. This thread was part of what pushed the discussion. The change will ship in an upcoming release.
> 
> 
> 
> Thanks you!

Thank you for being open and doing the right thing.

RDeluxe commented:
> Hey @ajnart @RDeluxe — I hear the frustration, and I want to be transparent about where we landed.
> 
> During our Q3 planning today, we discussed the concurrent builds feature specifically. We've decided to make it fully open to everyone, with no Enterprise-only limits. We see concurrent builds as a core platform capability, and core capabilities belong in the open-source version — that's the open core model we want to follow.
> 
> Balancing sustainability with keeping Dokploy genuinely useful as OSS isn't always easy, and we won't always get the line right on the first try. This thread was part of what pushed the discussion. The change will ship in an upcoming release.
> 
> Thanks you!

Thank you for listening to the community. More time for us to contribute here.
